internal/core/src/common/Utils.h

```diff
@@ -166,4 +173,198 @@ MatchKnowhereError(knowhere::Status status) {
+            PanicInfo("unsupported");
+        }
+    }
+}
+
+inline ssize_t
+WriteFieldData(int fd, DataType data_type, const DataArray* data, size_t size) {
+    switch (data_type) {
+        case DataType::BOOL: {
+            return write(fd, data->scalars().bool_data().data().data(), size);
+        }
+        case DataType::INT8: {
+            auto src_data = data->scalars().int_data().data();
+            std::vector<int8_t> data_raw(src_data.size());
+            std::copy_n(src_data.data(), src_data.size(), data_raw.data());
+            return write(fd, data_raw.data(), size);
+        }
+        case DataType::INT16: {
+            auto src_data = data->scalars().int_data().data();
+            std::vector<int16_t> data_raw(src_data.size());
+            std::copy_n(src_data.data(), src_data.size(), data_raw.data());
+            return write(fd, data_raw.data(), size);
+        }
+        case DataType::INT32: {
+            return write(fd, data->scalars().int_data().data().data(), size);
+        }
+        case DataType::INT64: {
+            return write(fd, data->scalars().long_data().data().data(), size);
+        }
+        case DataType::FLOAT: {
+            return write(fd, data->scalars().float_data().data().data(), size);
+        }
+        case DataType::DOUBLE: {
+            return write(fd, data->scalars().double_data().data().data(), size);
+        }
+        case DataType::VARCHAR: {
+            auto begin = data->scalars().string_data().data().begin();
+            auto end = data->scalars().string_data().data().end();
+
+            ssize_t total_written{0};
+            while (begin != end) {
+                ssize_t written = write(fd, begin->data(), begin->size());
+                if (written < begin->size()) {
+                    break;
+                }
+                total_written += written;
+                begin++;
+            }
+            return total_written;
+        }
+        case DataType::VECTOR_FLOAT:
+            return write(fd, data->vectors().float_vector().data().data(), size);
+
+        case DataType::VECTOR_BINARY:
+            return write(fd, data->vectors().binary_vector().data(), size);
+
+        default: {
+            PanicInfo("unsupported");
+        }
+    }
+}
+
+// CreateMap creates a memory mapping,
+// if mmap enabled, this writes field data to disk and create a map to the file,
+// otherwise this just alloc memory
+inline void*
+CreateMap(int64_t segment_id, const FieldMeta& field_meta, const LoadFieldDataInfo& info) {
+    static int mmap_flags = MAP_SHARED;
+#ifdef MAP_POPULATE
+    // macOS doesn't support MAP_POPULATE
+    mmap_flags |= MAP_POPULATE;
+#endif
+    // Allocate memory
+    if (info.mmap_dir_path == nullptr) {
+        auto data_type = field_meta.get_data_type();
+        auto data_size = field_meta.get_sizeof() * info.row_count;
+        size_t size = datatype_is_string(data_type) ? GetStringsSize(info.field_data) : data_size;
```

Comment:
I've seen many datatype_is_string check
Should we change to is_variable_type for futher extension